### PR TITLE
CART-89 test: Update to dual interface server

### PR DESCRIPTION
**Pull request by frostedcmos**
- Add -f [filename] option to transfer file contents during ping check
- Fix -d option in order to proper pass domain
- Kill both servers on any server failure

Signed-off-by: Alexander Oganezov <alexander.a.oganezov@intel.com>

### CODE REVIEW

File: src/tests/ftest/cart/dual_iface_server.c [hunk 1]
```diff
@@ -74,6 +84,35 @@ struct crt_proto_format my_proto_fmt = {
...
 
+static int
+bulk_transfer_done_cb(const struct crt_bulk_cb_info *info)
+{
+	void	*buff;
+	int	rc;
+
+	if (info->bci_rc != 0) {
+		D_ERROR("Bulk transfer failed with rc=%d\n", info->bci_rc);
+		error_exit();
```
Comment (SABollinger) [inline src/tests/ftest/cart/dual_iface_server.c:95]: On error, your destination buffer is not freed.

File: src/tests/ftest/cart/dual_iface_server.c [hunk 2]
```diff
@@ -74,6 +84,35 @@ struct crt_proto_format my_proto_fmt = {
...
+	}
+
+	DBG_PRINT("Bulk transfer done\n");
+
+	rc = crt_reply_send(info->bci_bulk_desc->bd_rpc);
+	if (rc != 0) {
+		D_ERROR("Failed to send response\n");
+		error_exit();
```
Comment (SABollinger) [inline src/tests/ftest/cart/dual_iface_server.c:103]: Same comment as above.

File: src/tests/ftest/cart/dual_iface_server.c [hunk 3]
```diff
@@ -109,8 +148,48 @@ handler_ping(crt_rpc_t *rpc)
...
+		bulk_desc.bd_remote_hdl = input->bulk_hdl;
+		bulk_desc.bd_remote_off = 0;
+		bulk_desc.bd_local_hdl = dst_bulk;
+		bulk_desc.bd_local_off = 0;
+		bulk_desc.bd_len = input->file_size;
+		rc = crt_bulk_transfer(&bulk_desc, bulk_transfer_done_cb,
+				       dst, NULL);
+		if (rc != 0) {
+			D_ERROR("transfer failed; rc=%d\n", rc);
+			error_exit();
```
Comment (SABollinger) [inline src/tests/ftest/cart/dual_iface_server.c:183]: Free buffer on error.  What about rpc structure?

File: src/tests/ftest/cart/dual_iface_server.c [hunk 4]
```diff
@@ -109,8 +148,48 @@ handler_ping(crt_rpc_t *rpc)
...
+		struct crt_bulk_desc	bulk_desc;
+		crt_bulk_t		dst_bulk;
+		char			*dst;
+		d_sg_list_t		sgl;
+
+		D_ALLOC_ARRAY(dst, input->file_size);
+
+		rc = d_sgl_init(&sgl, 1);
+		if (rc != 0)
+			error_exit();
```
Comment (SABollinger) [inline src/tests/ftest/cart/dual_iface_server.c:161]: Free buffer

File: src/tests/ftest/cart/dual_iface_server.c [hunk 5]
```diff
@@ -167,63 +252,64 @@ server_main(d_rank_t my_rank, const char *str_port, const char *str_interface,
...
-		assert(0);
+		error_exit();
 	}
 
 	rc = crt_init("server_grp", CRT_FLAG_BIT_SERVER |
 		      CRT_FLAG_BIT_AUTO_SWIM_DISABLE);
 	if (rc != 0) {
 		D_ERROR("crt_init() failed; rc=%d\n", rc);
-		assert(0);
+		error_exit();
```
Comment (SABollinger) [inline src/tests/ftest/cart/dual_iface_server.c:270]: Free semaphore on error.

**Comment by SABollinger** [comment]
There are lots of cleanup issues for failure cases: de-allocated memory buffers, open file descriptors, rpc structures, etc.  
Suggest a common cleanup section rather than duplicating each instance.

**Comment by frostedcmos** [comment]
> There are lots of cleanup issues for failure cases: de-allocated memory buffers, open file descriptors, rpc structures, etc.
> Suggest a common cleanup section rather than duplicating each instance.

This is a manual sample written such that we assert and kill a second server on any error. Since this is not intended to be running ever under the valgrind it does not require cleanups on errors

**Review by jolivier23 — APPROVED** [review]

**Review by SABollinger — APPROVED** [review]

**Pull request merged**